File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/randomized/P380_InsertDeleteGetRandom.java
package com.dsa.leetcode.labuladong.array.randomized;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class P380_InsertDeleteGetRandom {

    static class RandomizedSet {
        private List<Integer> nums;

        private Map<Integer, Integer> indices;

        private Random random;
        public RandomizedSet() {
            nums = new ArrayList<>();
            indices = new HashMap<>();
            random = new Random();
        }

        public boolean insert(int val) {
            if (indices.containsKey(val)) {
                return false;
            }
            indices.put(val, nums.size());
            nums.add(val);
            return true;
        }

        public boolean remove(int val) {
            if (!indices.containsKey(val)) {
                return false;
            }
            int index = indices.get(val);
            int last = nums.get(nums.size() - 1);
            nums.set(index, last);
            nums.remove(nums.size() - 1);
            indices.put(last, index);
            indices.remove(val);
            return true;
        }

        public int getRandom() {
            int i = random.nextInt(nums.size());
            return nums.get(i);
        }
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/array/randomized/P380_InsertDeleteGetRandom.py
# author: tobebetter9527
# since: 2024/2/20 8:40
import random


class RandomizedSet:

    def __init__(self):
        self.nums = []
        self.indices = {}

    def insert(self, val: int) -> bool:
        if val in self.indices:
            return False
        self.indices[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.indices:
            return False
        index = self.indices[val]
        lastVal = self.nums[-1]
        self.nums[index] = lastVal
        self.indices[lastVal] = index
        self.nums.pop()
        del self.indices[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.nums)
